### ISSUE TITLE
Opentracing not tracing graphql traces

### ISSUE THREAD

**Issue by F21** [neelance/graphql-go#53]
I have a JSON grpc server backed by this graphql library. The grpc server is instrumented using the `grpc-opentracing` package and I am using zipkin as the tracer. In the zipkin ui, I can see all the traces for the GRPC span, but I do not see any child graphql spans.

At the moment, I have only tested my mutations and have not tested any queries yet. The resolvers look like this:

```go
func (r *Resolver) DoSomething(args *struct {
	Input *DoSomethingInput
}) (*doSomethingPayload, error) {

}
```

I've added the following to https://github.com/neelance/graphql-go/blob/master/graphql.go#L86:
```go
sp := opentracing.SpanFromContext(ctx)
fmt.Println(sp)
```

It does find a span, but I see nothing in the zipkin ui:

```go
&{0xc4203fc310 <nil> {0 0} {{{0 21600196140346085} 1322274615697199075 true map[] <nil> 8 true} /eventstore.GraphQL/Query {63621715428 560669386 0x22de200} -1 map[span.kind:server component:gRPC] []} 0 <nil>}
```

**Comment by bsr203** [comment]
It does work for me but as the Readme says ` Fields are marked as trivial if they have no context.Context argument, no field arguments and no error result. You can use this to avoid unnecessary clutter in the trace:` Do you think that could be it?

**Comment by F21** [comment]
Which tracer are you using? I am using zipkin.

Having said that, my handlers have field arguments and an error result, so it shouldn't be marked as trivial. However, even if it was marked as trivial, I should still see the trace because I am not doing any filtering.

**Comment by bsr203** [comment]
I don't use zipkin, but just collect them in memory and just display per request (traceid is matched again client request). I them trivially display trace, span and log in my frontend in case any error.

**Comment by F21** [comment]
Do you mind sharing some sample code of how you are doing that in memory? I will give that a go tomorrow (getting late here) to see if the spans show up.

**Comment by bsr203** [comment]
I am just starting my day here. will look into it later.

**Comment by F21** [comment]
Awesome. Thanks! 😄

**Comment by bsr203** [comment]
I keep local structures for `Span`, Log` to map to data I am interested in, and also indices of span in trace id etc for serving. I will use zipkin etc later, but now just want to see the error during development.


```
func main() {
	opentracing.InitGlobalTracer(monit.NewTracer())
  ...

  }

func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
....
	span, c := opentracing.StartSpanFromContext(ctx, "HTTP")
	defer span.Finish()
span.LogFields(
		log.Object("header", getHeaderFromReq(r)),
		log.String("query", params.Query),
		log.String("operationName", params.OperationName),
		log.Object("variables", params.Variables),
	)
}


pkg monit
-----
  import (
   ...
  	basictracer "github.com/opentracing/basictracer-go"
  	opentracing "github.com/opentracing/opentracing-go"
  )

  var (
  	rec *basictracer.InMemorySpanRecorder
  )

  func NewTracer() opentracing.Tracer {
  	return basictracer.New(rec)
  }


  func SpansOnTrace(traceID string) ([]Span, error) {
  	raws := rec.GetSpans()
  	var pp []Span
  	for _, r := range raws {
  		s, err := mapFromRaw(r)
  		if err != nil {
  			return nil, err
  		}
  		pp = append(pp, s)
  	}
  	return pp, nil
  }

  ..

  func mapFromRaw(r basictracer.RawSpan) (Span, error) {
  	l, err := mapLogs(r.Logs)
  	if err != nil {
  		return Span{}, err
  	}
  	return Span{
  		TraceID:      strconv.FormatUint(r.Context.TraceID, 10),
  		SpanID:       strconv.FormatUint(r.Context.SpanID, 10),
  		Sampled:      r.Context.Sampled,
  		ParentSpanID: strconv.FormatUint(r.ParentSpanID, 10),
  		Operation:    r.Operation,
  		Start:        r.Start.UTC().Format(time.RFC3339),
  		Duration:     int32(r.Duration.Nanoseconds()), // TODO may need to divide first to get better precision
  		Logs:         l,
  	}, nil
  }

```

**Comment by F21** [comment]
@bsr203 Thanks! Your code was really useful.

I did some debugging over the weekend and found the source of the problem. In my case, the grpc opentracing intercept requires me to explicitly pass in a tracer. I was under the assumption that once a tracer was passed in at the top most entrypoint of the app, that would be it. However, `StartSpanFromContext` which is called by `graphql-go` will start a span from the global tracer (which is a noop tracer if we do not set a global tracer).

In my case, I set a global tracer using `opentracing.SetGlobalTracer()` at the earliest part of my code and replaced method arguments that required an explicit tracer to `opentracing.GlobalTracer()`.
